### ISSUE TITLE
Circular reference in cache

### ISSUE THREAD

**Issue by MartinMystikJonas** [efabrica-team/phpstan-latte#51]
```
PHP Warning:  var_export does not handle circular references in phar:///app/tools/devstack/phpstan/phpstan/phpstan.phar/src/Analyser/ResultCache/ResultCacheManager.php on line 551
```

Will investigate further

**Comment by MartinMystikJonas** [comment]
Caused by this code in ResultCacheManager.php:
```php
FileWriter::write(
  $file, 
  sprintf($template, var_export($lastFullAnalysisTime, \true),
  var_export($meta, \true),
  var_export($this->getProjectExtensionFiles($projectConfigArray, $dependencies), \true),
  var_export($errors, \true), var_export($collectedData, \true),
  var_export($invertedDependencies, \true),
  var_export($exportedNodes, \true))
);
```

**Comment by lulco** [comment]
It seems in tests MemoryCache is used (same as in debug mode) but in real run there is some file cache. We have to use it in tests too to replicate issue

**Comment by MartinMystikJonas** [comment]
This will show in debug mode. In production mode it either does not happen or is not shown in output.

**Comment by MartinMystikJonas** [comment]
Does not happen when analysing just one file (one by one) but happens when analyzing whole folder

**Comment by MartinMystikJonas** [comment]
Narrowed to VariableCollector and use of Type in Variable. It seems that results cache have problem with stroing type object in Results cache

**Comment by lulco** [comment]
Yes, that's why I've store just string in cache and than recreate type in variable finder

**Comment by lulco** [comment]
Hmm var export is called here, in my case it was json encode so we have to implement __set_state too or something similar

**Comment by MartinMystikJonas** [comment]
Or do not use Variable at all in collectors but store just strings in ValueObjects. ANd then recreate from it. Just extract type nad variable name in constructor and store it in provate properties.

**Comment by MartinMystikJonas** [comment]
```php
final class CollectedVariable implements JsonSerializable
{
    private string $className;

    private string $methodName;

    private string $variableName;

    private string $variableType;

    public function __construct(string $className, string $methodName, Variable $variable)
    {
        $this->className = $className;
        $this->methodName = $methodName;
        $this->variableName = $variable->getName();
        $this->variableType = $variable->getTypeAsString();
    }
```

**Comment by MartinMystikJonas** [comment]
And then reconstruct Variable with Type when Collector data are loaded. Simply avoid objects in properties of collector Value Objects

**Comment by MartinMystikJonas** [comment]
Should I prepare PR with these string only properties?

**Comment by lulco** [comment]
Yes this will be probably solution. Use arrays for all collectors and build objects then. I just don’t get it why I don’t have this problem in my projects

**Comment by MartinMystikJonas** [comment]
What version of phpstan do you have? Did you somehow disabled results cache?

**Comment by MartinMystikJonas** [comment]
PR here https://github.com/efabrica-team/phpstan-latte/pull/54

**Comment by MartinMystikJonas** [comment]
I reused Vakue object but I added parameter to force it to forget variable/component. Then if it is forgotten it is recreated in Finder

**Comment by lulco** [comment]
> What version of phpstan do you have? Did you somehow disabled results cache?

1.9.1 I think. I didn’t disable it. Nevermind, with arrays it will work great. I should follow docs in the first place :) I just wanted to have some strict types, not arrays :)

**Comment by lulco** [comment]
Fixed by #54
